File: image_search/search_vary_numtestimages.m
function [rank_baseline, rank_specificity] = search_vary_numtestimages(dataset, method)

addpath('../io/'); addpath('utils/');

[~, ~, s, ~, ~, ~] = load_search_parameters(dataset);

if strcmpi(dataset, 'clipart')
    s = s(51:end, 51:end); % XXX: Leave out seed images
end

if strcmpi(method, 'groundtruth')
    load('../../data/specificity_alldatasets.mat');
    eval(['y = specificity.' dataset '.B0;']);
    eval(['z = specificity.' dataset '.B1;']);
elseif strcmpi(method, 'predicted_1000fold')
    load(['../../data/predict_search/' dataset '/predicted_specificity_1000fold.mat']);
    y = y_pred(1, :); z = z_pred(1, :);
end

n_images = length(y);
test_size = 100:10:n_images;
rank_baseline = cell(length(test_size));
rank_specificity = cell(length(test_size));

for idx=1:length(test_size)

    fprintf('\nTest size = %d', test_size(idx));
    test_idx = randsample(n_images, test_size(idx));
    s_test = s(test_idx, test_idx);
    
    % BASELINE
    rank_b = baseline_search(s_test);
    rank_baseline{idx} = rank_b;

    % GROUND TRUTH / PREDICTED SPECIFICITY
    rank_s = specificity_search(s_test, y(test_idx), z(test_idx));
    rank_specificity{idx} = rank_s;

    % SAVE RESULTS
    if strcmpi(method, 'groundtruth')
        save(['../../data/predict_search/' dataset '/groundtruth_logistic.mat'], ...
              'rank_specificity', 'rank_baseline', 'test_size');
    elseif strcmpi(method, 'predicted_1000fold')
        save(['../../data/predict_search/' dataset '/predicted_logistic_1000fold.mat'], ...
              'rank_specificity', 'rank_baseline', 'test_size');
    end
end

end

File: image_search/search_vary_numtestimages.py
from operator import itemgetter
import numpy as np
import scipy.io as io
from joblib import Parallel, delayed

from utils.search import (search_specificity, search_baseline)


def vary_test_size(s, y, z, test_size):
    n_images = len(y)
    # note: this will give different result from matlab b/c of
    # different random number generators
    test_idx = np.random.choice(n_images, test_size, replace=False)

    #s_test = np.asarray([s[i][test_idx] for i in test_idx])
    s_test = np.asarray([itemgetter(test_idx)(s_t) for
                         s_t in itemgetter(test_idx)(s)])
    rank_b = search_baseline(s_test)
    rank_s = search_specificity(s_test, y[test_idx], z[test_idx])

    return rank_b, rank_s, test_idx


if __name__ == '__main__':

    dataset = 'pascal'

    mat1 = io.loadmat('search_parameters_pascal.mat')
    s = mat1['s']

    mat2 = io.loadmat('../../data/predict_search/%s/'
                      'predicted_specificity_1000fold.mat'
                      % dataset)
    y, z = mat2['y_pred'][0, :], mat2['z_pred'][0, :]
    n_images = len(y)

    parallel, p_func = Parallel(n_jobs=6, verbose=30), delayed(vary_test_size)
    r = parallel(p_func(s, y, z, test_size) for test_size in
                 range(100, n_images + 1, 10))
    rank_b, rank_s, test_idx = zip(*r)
    savedict = {'rank_baseline': rank_b, 'rank_specificity': rank_s,
                'test_idx': test_idx, 'test_size': range(100, n_images + 1, 10)}
    io.savemat('../../data/predict_search/%s/'
               'predicted_logistic_1000fold_python.mat' % dataset, savedict)
